File: Code/Automat.java
import AlgoTools.IO;
/**
 * Endlicher Automat zur Überprüfung, ob eine beliebig
 * lange Dualzahl durch 3 teilbar ist.
 */
public class Automat {

  public static void main(String[] args) {

    int[][] delta = { // Überführungsfunktion
      {0, 1},       
      {2, 0},
      {1, 2}
    };
    int s = 0;                     // Startzustand
    char[] zeile = IO.readChars(); // Eingabezeichen

    // Überführungsfunktion zeichenweise anwenden
    for (int i = 0; i < zeile.length; i++)
      s = delta[s][zeile[i] - '0'];

    // Endzustand prüfen
    if (s == 0)
      IO.println("Eingabe ist durch 3 teilbar.");
    else
      IO.println("Eingabe ist NICHT durch 3 teilbar.");
  }

}

File: Code/automat.pas
program Automat;
  (* Endlicher Automat mit delta: Zustand x Eingabe -> Zustand.
   *  - Ueberprueft, ob ein eingegebener Binaerstring durch 3
   *    teilbar ist.
   *  - Syntaktisch korrekte Eingabe wird vorausgesetzt.
   *  - Zeichen '0' und '1' werden umgewandelt in Integer 0 und 1.
   *
   * 2017-01-03 (kb)
   *)
var
 c: char;                                   (* Eingabezeichen     *)
 d: array [0..2,0..1] of integer = ((0,1),  (* Ueberfuehrungsfkt. *)
                                    (2,0),
                                    (1,2));
 s:  integer;                               (* aktueller Zustand  *)
begin
  s:=0;                                     (* Startzustand       *)
  while not Eoln do begin
    read(c);                                (* naechstes Zeichen  *)
    s:=d[s,Ord(c)-Ord('0')]                 (* delta anwenden     *)
  end;
  if (s=0) then                             (* Endzustand pruefen *)
    writeLn('Eingabe ist durch 3 teilbar.')
  else
    writeLn('Eingabe ist NICHT durch 3 teilbar.')
end.
